---
title: "Class 17:"
author: "Katherine Lim (A15900881)"
format: pdf
---

# 1. Investigating pertussis cases by year

Pertussis or whooping cough is a highly contagious lung infection caused by the bacteria *B. pertussis*.

The CDC tracks reported cases in the US since the 1920s. 

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
cdc <- data.frame(
        Year = c(1922L,1923L,1924L,1925L,
        1926L,1927L,1928L,1929L,1930L,1931L,
        1932L,1933L,1934L,1935L,1936L,
        1937L,1938L,1939L,1940L,1941L,1942L,
        1943L,1944L,1945L,1946L,1947L,
        1948L,1949L,1950L,1951L,1952L,
        1953L,1954L,1955L,1956L,1957L,1958L,
        1959L,1960L,1961L,1962L,1963L,
        1964L,1965L,1966L,1967L,1968L,1969L,
        1970L,1971L,1972L,1973L,1974L,
        1975L,1976L,1977L,1978L,1979L,1980L,
        1981L,1982L,1983L,1984L,1985L,
        1986L,1987L,1988L,1989L,1990L,
        1991L,1992L,1993L,1994L,1995L,1996L,
        1997L,1998L,1999L,2000L,2001L,
        2002L,2003L,2004L,2005L,2006L,2007L,
        2008L,2009L,2010L,2011L,2012L,
        2013L,2014L,2015L,2016L,2017L,2018L,
        2019L,2020L,2021L),
        
        Cases = c(107473,164191,165418,152003,
        202210,181411,161799,197371,
        166914,172559,215343,179135,265269,
        180518,147237,214652,227319,103188,
        183866,222202,191383,191890,109873,
        133792,109860,156517,74715,69479,
        120718,68687,45030,37129,60886,
        62786,31732,28295,32148,40005,
        14809,11468,17749,17135,13005,6799,
        7717,9718,4810,3285,4249,3036,
        3287,1759,2402,1738,1010,2177,2063,
        1623,1730,1248,1895,2463,2276,
        3589,4195,2823,3450,4157,4570,
        2719,4083,6586,4617,5137,7796,6564,
        7405,7298,7867,7580,9771,11647,
        25827,25616,15632,10454,13278,
        16858,27550,18719,48277,28639,32971,
        20762,17972,18975,15609,18617,
        6124,2116)
       )
```

```{r}
library(ggplot2)

ggplot(cdc) +
  aes(Year, Cases) +
  geom_point() +
  geom_line() +
  labs(x = "Year", Y = "Cases") +
  geom_vline(xintercept = 1942, color = "blue") +
  geom_vline(xintercept = 1980, color = "grey", linetype = 2) +
  geom_vline(xintercept = 1995, color = "red") +
  scale_y_continuous(labels = scales::comma)
```

The first big "whole-cell" pertussis vaccine program started in 1942.

# 2. A tale of two vaccines

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine.

```{r}
library(ggplot2)

ggplot(cdc) +
  aes(Year, Cases) +
  geom_point() +
  geom_line() +
  labs(x = "Year", Y = "Cases") +
  geom_vline(xintercept = 1946, color = "blue") +
  geom_vline(xintercept = 1996, color = "red") +
  scale_y_continuous(labels = scales::comma)
```

Something big is happening with pertussis and big outbreaks are once again a major public health concern. 

One of the main hypotheses for the increasing case numbers is waning vaccine efficiency with the newer aP vaccine.

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

There is a slight increase in cases following the switch to the aP vaccine.

# 3. Exploring CMI-PB data

Enter the CMI-PB project, which is studying this problem on a large scale. Let's see what data they have.

Their data is available in JSON format ("key:value" pair style). We will use the "jsonlight" package to read their data.

```{r}
library(jsonlite)

subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE)
head(subject, 3)
```

> Q4. How may aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc.)?

```{r}
table(subject$race, subject$biological_sex)
```

> Q7. Determine (i) the average age of wP individuals, (ii) the average age of aP individuals, and (iii) are they significantly different?

```{r}
library(lubridate)
library(dplyr)

subject$age <- today() - ymd(subject$year_of_birth)

# (i) the average age of wP individuals
wp <- subject %>% filter(infancy_vac == "wP")

round( summary( time_length( wp$age, "years" ) ) )

# (ii) the average age of aP individuals
ap <- subject %>% filter(infancy_vac == "aP")

round( summary( time_length( ap$age, "years" ) ) )

# (iii) are they significantly different
t_test <- t.test(wp$age, ap$age)
print(t_test)
```

> Q8. Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

> Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill = as.factor(infancy_vac)) +
  geom_histogram(show.legend = FALSE) +
  facet_wrap(vars(infancy_vac), nrow = 2) 
```

Yes.

Now let's read some more database tables from CMI-PB.

```{r}
specimen <- read_json("http://cmi-pb.org/api/specimen", simplifyVector = TRUE)
head(subject)
```

I want to join" the `subject` and `specimen` tables together. We can use the `dplyr` package for this. 

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details.

```{r}
library(dplyr)

meta <- inner_join(subject, specimen)
head(meta)
```

> Q10. Now using the same procedure join ab with meta data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
ab <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE)

abdata <- inner_join(meta, ab)
head(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

> Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```

There are way less visit 8 specimens because the project is still ongoing and we do not have all the data for those individuals yet.

# 4. Examine IgG1 Ab titer levels

We will use the `filter()` function from dplyr to focus on just IgG1 isotype and visits 1 to 7.

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit != 8)
head(ig1)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow = 2)
```

> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

The FIM 2/3 (fimbrial protein), PT (pertussis toxin), FHA (filamentous hemagglitinin). 

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. 

```{r}
filter(ig1, antigen == "Measles") %>%
  ggplot() +
  aes(MFI, col = infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(ig1, antigen == "FIM2/3") %>%
  ggplot() +
  aes(MFI, col = infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

> Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?

FIM2/3 levels  rise over time and exceed those of Measles. They also appear to peak at visit 5 and then decline.

> Q17. Do you see any clear difference in aP vs. wP responses?

wP and aP responses follow a similar trend. 

# 5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 

ssrna <- inner_join(rna, meta)
```

> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group = subject_id) +
  geom_point() +
  geom_line(alpha = 0.2)
```

> Q19. What do you notice about the expression of this gene?

The expression of this gene is at its maximum at visit 4.








